Doc/extending/newtypes_tutorial.rst

```diff
@@ -0,0 +1,865 @@
+
+   >>> import noddy
+   >>> mynoddy = noddy.Noddy()
+
+That's it!  All that remains is to build it; put the above code in a file called
+:file:`noddy.c` and ::
+
+   from distutils.core import setup, Extension
+   setup(name="noddy", version="1.0",
+         ext_modules=[Extension("noddy", ["noddy.c"])])
+
+in a file called :file:`setup.py`; then typing
+
+.. code-block:: shell-session
+
+   $ python setup.py build
+
+at a shell should produce a file :file:`noddy.so` in a subdirectory; move to
+that directory and fire up Python --- you should be able to ``import noddy`` and
+play around with Noddy objects.
+
+That wasn't so hard, was it?
+
+Of course, the current Noddy type is pretty uninteresting. It has no data and
+doesn't do anything. It can't even be subclassed.
+
+
+Adding data and methods to the Basic example
+============================================
+
+Let's extend the basic example to add some data and methods.  Let's also make
+the type usable as a base class. We'll create a new module, :mod:`noddy2` that
+adds these capabilities:
+
+.. literalinclude:: ../includes/noddy2.c
+
+
+This version of the module has a number of changes.
+
+We've added an extra include::
+
+   #include <structmember.h>
+
+This include provides declarations that we use to handle attributes, as
+described a bit later.
+
+The  :class:`Noddy` type now has three data attributes in its C struct,
+*first*, *last*, and *number*.  The *first* and *last* variables are Python
+strings containing first and last names.  The *number* attribute is a C integer.
+
+The object structure is updated accordingly::
+
+   typedef struct {
+       PyObject_HEAD
+       PyObject *first; /* first name */
+       PyObject *last;  /* last name */
+       int number;
+   } NoddyObject;
+
+Because we now have data to manage, we have to be more careful about object
+allocation and deallocation.  At a minimum, we need a deallocation method::
+
+   static void
+   Noddy_dealloc(NoddyObject *self)
+   {
+       Py_XDECREF(self->first);
+       Py_XDECREF(self->last);
+       Py_TYPE(self)->tp_free((PyObject *) self);
+   }
+
+which is assigned to the :c:member:`~PyTypeObject.tp_dealloc` member::
+
+    .tp_dealloc = (destructor) Noddy_dealloc,
+
+This method first clears the reference counts of the two Python attributes.
+:c:func:`Py_XDECREF` correctly handles the case where its argument is
+*NULL* (which might happen here if ``tp_new`` failed midway).  It then
+calls the :c:member:`~PyTypeObject.tp_free` member of the object's type
+(computed by ``Py_TYPE(self)``) to free the object's memory.  Note that
+the object's type might not be :class:`NoddyType`, because the object may
+be an instance of a subclass.
+
+.. note::
+   The explicit cast to ``destructor`` above is needed because we defined
+   ``Noddy_dealloc`` to take a ``NoddyObject *`` argument, but the ``tp_dealloc``
+   function pointer expects to receive a ``PyObject *`` argument.  Otherwise,
```

Comment:
It's an idiom that is safely used in many parts of CPython. I don't think it's "undefined" in C to do that (a pointer is a pointer, regardless of what it points to), unlike perhaps C++. Do you have a reference?

The alternative needs either many casts or using an additional variable inside the function, both of which are more cumbersome.

Reply:
It's compllicated, so I could be wrong, but C99 says "If a converted pointer is used to call a function whose type is not compatible with the pointed-to type, the behavior is undefined" ([sect. 6.3.2.3 par. 8](https://port70.net/~nsz/c/c99/n1256.html#6.3.2.3p8)). The function types are not compatible because "For two function types to be compatible, . . . corresponding parameters shall have compatible types" ([sect. 6.7.5.3 par. 15](https://port70.net/~nsz/c/c99/n1256.html#6.7.5.3p15)). The parameters are not compatible because they point to incompatible structures (NoddyObject and PyObject): "For two pointer types to be compatible, . . . both shall be pointers to compatible types" ([sect. 6.5.7.2 par. 2](https://port70.net/~nsz/c/c99/n1256.html#6.7.5.1p2)).

Yes, Python's own code does this function pointer casting a lot itself with success, but it never seemed right to me, and I don't think it is good to claim the cast is "needed" when there is an alternative.

There are a couple of recent bug reports about GCC 8 producing new warnings about explicit function pointer casts, e.g. <https://bugs.python.org/issue33012> about METH_NOARGS and PyCFunction.

Reply:
bpo-33012 is about wrong number of arguments, right?

Reply:
I think you are right. Looking closer, the bug reports seem to be about GCC's "-Wcast-function-type" warning, and one of the documented exceptions to that warning is "Any parameter of pointer-type matches any other pointer-type."